src/server_manager/electron_app/digitalocean_oauth.ts

```diff
@@ -0,0 +1,231 @@
+        }
+      }
+      server.close();
+      reject(error);
+    });
+    server.listen({host: 'localhost', port: portList[portIdx], exclusive: true});
+  });
+}
+
+// See https://developers.digitalocean.com/documentation/v2/#get-user-information
+interface Account {
+  droplet_limit: number;
+  floating_ip_limit: number;
+  email: string;
+  uuid: string;
+  email_verified: boolean;
+  status: string;
+  status_message: string;
+}
+
+// Queries the DigitalOcean API for the user account information.
+function getAccount(accessToken: string): Promise<Account> {
+  return new Promise((resolve, reject) => {
+    request(
+        {
+          url: 'https://api.digitalocean.com/v2/account',
+          headers: {
+            'User-Agent': 'Outline Manager',
+            'Content-Type': 'application/json',
+            'Authorization': `Bearer ${accessToken}`
+          }
+        },
+        (error, response, body) => {
+          if (error) {
+            return reject(error);
+          }
+          const bodyJson: {account: Account} = JSON.parse(body);
+          return resolve(bodyJson.account as Account);
+        });
+  });
+}
+
+export interface OauthSession {
+  // The result of the OAuth session, with the authentication token.
+  result: Promise<string>;
+  // Checks if the OauthSession has been cancelled.
+  isCancelled(): boolean;
+  // Cancels the OAuth session. The result promise is automatically rejected
+  // and isCancelled() will return true.
+  cancel(): void;
+}
+
+const CLOSE_WINDOW_HTML = `<html><script>window.close()</script></html>`;
+
+// Runs the DigitalOcean oauth flow and returns the access token.
+// See https://developers.digitalocean.com/documentation/oauth/ for the OAuth API.
+export function runOauth(): OauthSession {
+  const secret = randomValueHex(16);
+
+  const app = express();
+  const server = http.createServer(app);
+  server.on('close', () => console.log('Oauth server closed'));
+
+  let isCancelled = false;
+  // Disable caching.
+  app.use((req, res, next) => {
+    res.header('Cache-Control', 'no-cache, no-store, must-revalidate');
+    next();
+  });
+  // Check for cancellation.
+  app.use((req, res, next) => {
+    if (isCancelled) {
+      res.status(503).send('Authentication cancelled');
+    } else {
+      next();
+    }
+  });
+
+  // This is the callback for the DigitalOcean callback. It serves Javascript that will
+  // extract the access token from the hash and post it back to our http server.
+  app.get('/', (request, response) => {
+    response.send(`<html>
+          <head><title>Authenticating...</title></head>
+          <body>
+              <form id="form" method="POST">
+                  <input id="params" type="hidden" name="params"></input>
+              </form>
+              <script>
+                  let params = new URLSearchParams(location.hash.substr(1));
+                  let form = document.getElementById("form");
+                  let targetUrl = params.get("state");
```

Comment:
This Javascript uses some features that aren't universally available:
1. `let` was added in IE 11 but is not in IE 10.  Consider sticking to `var`.
2. `URLSearchParams` still doesn't exist in IE 11.

Since this page opens in the user's default browser, I think we should avoid these features.

Reply:
Changed `let` to `var` and added a simple URL params parsing.